src/librustc_mir/transform/check_consts/validation.rs

```diff
@@ -0,0 +1,586 @@
+                    assert!(op == BinOp::Eq || op == BinOp::Ne ||
+                            op == BinOp::Le || op == BinOp::Lt ||
+                            op == BinOp::Ge || op == BinOp::Gt ||
+                            op == BinOp::Offset);
+
+
+                    self.check_op(ops::RawPtrComparison);
+                }
+            }
+
+            Rvalue::NullaryOp(NullOp::Box, _) => {
+                self.check_op(ops::HeapAllocation);
+            }
+        }
+    }
+
+    fn visit_place_base(
+        &mut self,
+        place_base: &PlaceBase<'tcx>,
+        context: PlaceContext,
+        location: Location,
+    ) {
+        trace!(
+            "visit_place_base: place_base={:?} context={:?} location={:?}",
+            place_base,
+            context,
+            location,
+        );
+        self.super_place_base(place_base, context, location);
+
+        match place_base {
+            PlaceBase::Local(_) => {}
+            PlaceBase::Static(box Static{ kind: StaticKind::Promoted(_, _), .. }) => {
+                bug!("Promotion must be run after const validation");
+            }
+
+            PlaceBase::Static(box Static{ kind: StaticKind::Static, def_id, .. }) => {
+                let is_thread_local = self.tcx.has_attr(*def_id, sym::thread_local);
+                if is_thread_local {
+                    self.check_op(ops::ThreadLocalAccess);
+                } else if self.mode == Mode::Static && context.is_mutating_use() {
+                    // this is not strictly necessary as miri will also bail out
+                    // For interior mutability we can't really catch this statically as that
+                    // goes through raw pointers and intermediate temporaries, so miri has
+                    // to catch this anyway
+
+                    self.tcx.sess.span_err(
+                        self.span,
+                        "cannot mutate statics in the initializer of another static",
+                    );
+                } else {
+                    self.check_op(ops::StaticAccess);
+                }
+            }
+        }
+    }
+
+    fn visit_assign(&mut self, dest: &Place<'tcx>, rvalue: &Rvalue<'tcx>, location: Location) {
+        trace!("visit_assign: dest={:?} rvalue={:?} location={:?}", dest, rvalue, location);
+
+        // Error on mutable borrows or shared borrows of values with interior mutability.
+        //
+        // This replicates the logic at the start of `assign` in the old const checker.  Note that
+        // it depends on `HasMutInterior` being set for mutable borrows as well as values with
+        // interior mutability.
+        if let Rvalue::Ref(_, kind, ref borrowed_place) = *rvalue {
+            let rvalue_has_mut_interior = HasMutInterior::in_rvalue(
+                &self.item,
+                self.qualifs.has_mut_interior.get(),
+                rvalue,
+            );
+
+            if rvalue_has_mut_interior {
```

Comment:
Like I mentioned elsewhere, you can use `HasMutInterior` for shared borrows and then handle mutable borrows here specially, without `HasMutInterior` having to understand `Rvalue::Ref` *at all*.

Reply:
See [above](https://github.com/rust-lang/rust/pull/64470#discussion_r328272600).